[#13] 친구추가 기능 구현

msugo1 opened the pull request:
## Enum 클래스 관련

### FriendStatus 클래스 추가
- `친구 상태`에 대한 상수, 카테고리 값을 가지고 있는 Enum 클래스

- `TypeHandler`를 사용해서 자동으로 DB 저장시에는 정수 값을, 불러올 때는 상수
  값을 가져올 수 있도록 함

## Controller 계층관련

### FriendController 클래스 추가

- 친구 추가 요청을 받을 `sendFriendRequest` 메소드 추가

## Service 계층 관련

### FriendService 클래스 추가

- 요청에 대한 조건을 검사한 뒤 친구추가 요청을 보낼 `sendFriendRequest` 메소드 추가

  1. 요청을 보낼 사용자의 아이디가 `현재 사용자의 아이디와 같은지` 확인

  2. 요청을 보낼 사용자의 아이디가 `서비스 내에 실제로 존재하는지` 확인

  3. 친구 `status` 값 검사를 통해 `중복/부적절한 요청` 확인
  
  4. `문제 없을 시 친구추가` 요청

## Mapper 계층 관련

### FriendMapper 인터페이스 추가

- 보낸 친구 요청을 DB에 반영하기 위한 `sendFriendRequest` 메소드 추가

- DB에 저장된 친구 status 값을 불러오기 위한 `getFriendStatus` 메소드
  추가

### FriendMapper.xml 파일 추가

- 위의 메소드의 목적에 맞게 동작할 쿼리 문장을 담고 있는 mapper.xml 파일

## Exception 관련

### DuplicateFriendRequestException 클래스 추가

- `중복된 친구 추가 요청`에 대한 예외를 처리하기 위한 클래스

  : `DuplicateRequestException`을 재사용 하려고 했으나 ExceptionHandler가 `vague statement` 오류를 발생시켜 새로운 예외 클래스를 작성

### InvalidFriendRequestException 클래스 추가

- `부적절한 친구 추가 요청`에 대한 예외를 처리하기 위한 클래스

### 해당 예외들을 ControllerExceptionHandler에 매핑

- 위의 예외가 발생 시 예외 메시지와 함께 `Bad Request(400)`을 사용자에게
  리턴

## 테스트 관련

* FriendService 단위 테스트 작성

- FriendService의 `sendFriendRequest` 메소드에 대한 발생가능한 예외/성공 케이스에 대한
단위테스트 작성 및 구동

- Postman 을 활용해 예상 가능한 예외/성공 케이스에 대한 로직이 제대로 작동하는지 확인

Commit:
[#13] 친구추가 기능 구현

Enum 클래스 관련

* FriendStatus 클래스 추가
- `친구 상태`에 대한 상수, 카테고리 값을 가지고 있는 Enum 클래스
- `TypeHandler`를 사용해서 자동으로 DB 저장시에는 정수 값을, 불러올 때는 상수
  값을 가져올 수 있도록 함

Controller 계층관련

* FriendController 클래스 추가

- 친구 추가 요청을 받을 `sendFriendRequest` 메소드 추가

Service 계층 관련

* FriendService 클래스 추가

- 요청에 대한 조건을 검사한 뒤 친구추가 요청을 보낼 `sendFriendRequest` 메소드 추가

1. 요청을 보낼 사용자의 아이디가 `현재 사용자의 아이디와 같은지` 확인

2. 요청을 보낼 사용자의 아이디가 `서비스 내에 실제로 존재하는지` 확인

3. 친구 `status` 값 검사를 통해 중복/부적절한 요청 확인

4. 문제 없을 시 친구추가 요청

Mapper 계층 관련

* FriendMapper 인터페이스 추가

- 보낸 친구 요청을 DB에 반영하기 위한 `sendFriendRequest` 메소드 추가

- DB에 저장된 친구 status 값을 불러오기 위한 `getFriendStatus` 메소드
  추가

* FriendMapper.xml 파일 추가

- 위의 메소드의 목적에 맞게 동작할 쿼리 문장을 담고 있는 mapper.xml 파일

Exception 관련

* DuplicateFriendRequestException 클래스 추가

- `중복된 친구 추가 요청`에 대한 예외를 처리하기 위한 클래스
: `DuplicateRequestException`을 재사용 하려고 했으나
  ExceptionHandler가 `vague statement` 오류를 발생시켜 클래스를 하나
  다시 만듦

* InvalidFriendRequestException 클래스 추가

- `부적절한 친구 추가 요청`에 대한 예외를 처리하기 위한 클래스

* 해당 예외들을 ControllerExceptionHandler에 매핑

- 위의 예외가 발생 시 예외 메시지와 함께 `Bad Request(400)`을 사용자에게
  리턴

테스트 관련

* FriendService 단위 테스트 작성

- FriendService의 `sendFriendRequest` 메소드에 대한 발생가능한 예외/성공 케이스에 대한
단위테스트 작성 및 구동

- Postman 을 활용해 예상 가능한 예외/성공 케이스에 대한 로직이 제대로 작동하는지 확인

Commit:
[#13] 단위 테스트 @Display에 있는 오타 수정

msugo1 commented:
질문사항이 있어서 여기에 남깁니다.

1. 친구 추가를 할 때 사용자의 아이디가 실제로 존재하는 지 체크하는 과정이 필수일까요? 일반적으로 사용자의 프로파일 등을 보고 친구추가를 하는게 맞는데 혹시 몰라서 예외로 잡아뒀거든요. 필수가 아니면 DB에 대한 접근을 한 번 줄일 수 있어서 여쭤봅니다.

2.  사용자의 아이디가 존재하는지 확인하기 위해 새로운 메소드를 `UserService` 내에 생성하려다 해당 클래스 내에 `isUserIdDuplicate`가 사실은 똑같은 기능을 하는 것 같아서 대신 사용했는데요. 이름을 `doesUserIdExist` 쯤으로 바꾸면 사용자 아이디 `중복 체크`와 `사용자가 존재하는지 체크`하는 기능에 모두 적합해 질 것 같아서 이름을 바꾸려고 하다가 `친구 추가 기능 구현` 하는 브랜치에서 `UserService`를 건드리는 것이 맞나해서 일단 이름을 변경하지 않고 놔두었습니다. 

 -> 이름을 현재 브랜치에서 수정하는게 맞을까요? 아니면 `UserService`에 대한 리팩토링을 계속 고민중인데 나중에 리팩토링할 때 함께 고치는 것이 나을까요?

f-lab-dev commented:
1. 엉뚱한 값을 입력할 수 있으므로 필수적일 것 같습니다~ 엉뚱한 값을 넣어도 DB에 들어간다는것 자체가 치명적으로 보이네요~
2. 간단한거니 개발을 진행하는데 있어 큰 문제는 되지 않을 것 같습니다~ 여기서는 이름만 바꿔주고 나머지 `UserService` 리팩토링은 다른 브랜치에서 해주면 될 것 같아요~

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/FriendController.java:
```diff
@@ -0,0 +1,23 @@
+
+    private final FriendService friendService;
+
+    @PostMapping("/request/{anotherUserId}")
```
이런식으로도 표현해볼 수 있을 것 같아요~ `POST /friend-requests/from/{userId}/to/{userId}`

msugo1 replied:
`from - to -`가 정말 `URL` 이라 생각해서 처음에 `POST /friend-requests/from/{userId}/to/{targetId}` 로 변경하였는데, 포스트맨에서 테스트 해보니 앞에 `userId`가 존재하지 않아도 친구요청이 보내져 DB에 저장되는 문제가 있었습니다. 

그래서 친구 요청을 보내는 것은 어차피 로그인 한 현재유저일 것이라 생각하고 URL을 `POST /friend-requests/to/{targetId}` 로 대신 변경했습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/FriendController.java:
```diff
@@ -0,0 +1,23 @@
+    public ResponseEntity<Void> sendFriendRequest(@CurrentUser String userId, @PathVariable String anotherUserId) {
+        friendService.sendFriendRequest(userId, anotherUserId);
+
+        return HttpResponses.HTTP_RESPONSE_CREATED;
```
이거는 `@ResponseStatus`를 달아주면 생략이 가능해보입니다~

msugo1 replied:
일단 통일시켜놨다가 `UserService` 리팩토링 하면서 같이 바꾸려고 했는데 생각해보니 범위가 다르긴 하네요. `@ResponseStatus(HttpStatus.CREATED)`로 변경하고 리턴 타입은 생략해주었습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FriendService.java:
```diff
@@ -0,0 +1,51 @@
+    private final FriendMapper friendMapper;
+
+    public void sendFriendRequest(String userId, String anotherUserId) {
+        if (userId.equals(anotherUserId)) {
```
`userId`가 null일수도 있으므로 `StringUtils.equals` 같은 것을 사용해도 좋을 것 같네요~

msugo1 replied:
`Null-safe`한 `StringUtils.equals` 가 있다는 것을 몰랐네요! 변경하였습니다. 감사합니다.

Commit:
[#13] 피드백 반영

FriendController

* URL 변경
- 기존 친구추가URL을`POST friend-requests/to/{targetId}`로 변경

* `sendFriendRequest` 메소드 리턴타입 생략
- `@ResponseStatus(HttpStatus)` 형식의 annotation을 대신 붙여줌으로써
  리턴타입 생략

* 변수 이름 변경
- `anotherUserId` 를 좀 더 간결하게 `targetId` 로 변경

FriendService

* StringUtils.equals 도입
- 현재 사용자의 아이디와 친구 추가할 사용자의 아이디가 같은지 비교할 때
  현재 사용자의 아이디가 null인 경우를 감안하여 StringUtils.equals 사용

* 이에 필요한 의존성 추가
- `org.apache.commons`: 스프링 부트에서 관리하는 의존성으로 버전은 따로
  추가할 필요 없음

UserService

* isUserIdDuplicate 메소드 이름 변경
- FriendService에서 아이디를 존재 여부를 체크하는데 재사용할 수 있도록
`doesUserIdExist`로 메소드 이름 변경

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -32,7 +32,7 @@
     public ResponseEntity<Void> isUserIdDuplicate(@RequestParam String userId) {
 
-        if (userService.isUserIdDuplicate(userId)) {
+        if (userService.doesUserIdExist(userId)) {
```
일반적으로 boolean을 리턴하는 메소드 앞에는 `is`를 붙여줍니다~

f-lab-dev reviewed (approved):
코멘트 드린거만 고치면 머지해도 될 것 같네요~

Commit:
[#13] 피드백 반영

- boolean 메소드 이름 수정 does -> is
